File: KCD/HMKaisouLockCommand.m
//
//  HMKaisouLockCommand.m
//  KCD
//
//  Created by Hori,Masaki on 2014/10/10.
//  Copyright (c) 2014年 Hori,Masaki. All rights reserved.
//

#import "HMKaisouLockCommand.h"

#import "HMServerDataStore.h"
#import "HMKCSlotItemObject+Extensions.h"

@implementation HMKaisouLockCommand
+ (void)load
{
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		[HMJSONCommand registerClass:self];
	});
}

+ (BOOL)canExcuteAPI:(NSString *)api
{
	return [api isEqualToString:@"/kcsapi/api_req_kaisou/lock"];
}

- (void)execute
{
	NSDictionary *api_data = [self.json valueForKeyPath:self.dataKey];
	if(![api_data isKindOfClass:[NSDictionary class]]) {
		[self log:@"api_data is NOT NSDictionary."];
		return;
	}
	id slotitemId = self.arguments[@"api_slotitem_id"];
	
	HMServerDataStore *serverDataStore = [HMServerDataStore oneTimeEditor];	
	NSError *error = nil;
	NSArray<HMKCSlotItemObject *> *slotItems = [serverDataStore objectsWithEntityName:@"SlotItem"
																			 error:&error
																   predicateFormat:@"id = %ld", [slotitemId integerValue]];
	if(error) {
		[self log:@"Fetch error: %@", error];
		return;
	}
	if(slotItems.count == 0) {
		[self log:@"Could not find SlotItem number %@", slotitemId];
		return;
	}
	
	BOOL locked = [api_data[@"api_locked"] boolValue];
	slotItems[0].locked = @(locked);
}

@end

File: KCD/HMKaisouLockCommand.swift
//
//  HMKaisouLockCommand.swift
//  KCD
//
//  Created by Hori,Masaki on 2015/01/16.
//  Copyright (c) 2015年 Hori,Masaki. All rights reserved.
//

import Cocoa

class HMKaisouLockCommand: HMJSONCommand {
	required override init() {}
	
	override class func load() {
		// Mavericksにおいてload()のタイミングでNSArrayを生成することが出来ないため、遅延させる
		dispatch_after(dispatch_time(DISPATCH_TIME_NOW, Int64(1 * Double(NSEC_PER_MSEC))), dispatch_get_main_queue()) {
			HMJSONCommand.registerInstance(self())
		}
	}
	override class func canExcuteAPI(api: String) -> Bool {
		return api == "/kcsapi/api_req_kaisou/lock"
	}
	
	override func execute() {
		if let data = json[dataKey] as? NSDictionary {
			if let slotitemID = arguments["api_slotitem_id"] as? String {
				let store = HMServerDataStore.oneTimeEditor()
				var error: NSError? = nil
				let predicate = NSPredicate(format: "id = \(slotitemID)")
				let array = store.objectsWithEntityName("SlotItem",
					sortDescriptors: nil,
					predicate: predicate,
					error: &error)
				if error != nil {
					log("Feth error: \(error!)", argList: getVaList([]))
					return
				}
				if array.count == 0 {
					log("Could not find SlotItem number \(slotitemID)", argList: getVaList([]))
					return
				}
				if let locked = data["api_locked"] as? NSNumber {
					if let slotItem = array[0] as? HMKCSlotItemObject {
						slotItem.locked = locked.boolValue
					}
				}
			}
		} else {
			log("api_data is NOT NSDictionary", argList: getVaList([]))
		}
	}
}
